Ete4 #750

dengzq1234 opened the pull request:
this PR is to solve #750 

It includes the following features:

- Convert methods in `GTDBTaxa()` which involves meaningless numeric ids to internal methods, such as: 
    - `get_lineage_translator()` -> `_get_lineage_translator()`
    - `get_name_translator()` -> `_get_name_translator()`
    - `translate_to_names()` -> `_translate_to_names()`
- Convert input of `get_rank()`  from numeric id to string id in `GTDBTaxa()`, for example:
```
print(gtdb.get_rank(['c__Thorarchaeia', 'RS_GCF_001477695.1']))
{'c__Thorarchaeia': 'class', 'RS_GCF_001477695.1': 'subspecies'}
```

- Add flag `ignore_unclassified` in both `NCBITaxa()` and `GTDBTaxa()`, when `ignore_unclassified=True`, `annotate_tree()` will igore empty annotation of leaves

dengzq1234 committed:
add ignore unclassified in ncbi taxa

dengzq1234 committed:
update ncbi test

dengzq1234 committed:
allow ignore_unclassified in gtdb taxa module

dengzq1234 committed:
update db

dengzq1234 committed:
make get_lineage internal

jordibc commented on ete4/gtdb_taxonomy/gtdbquery.py:
```diff
@@ -146,13 +146,29 @@ def _translate_merged(self, all_taxids):
-    def get_rank(self, taxids):
+    def _dirty_id_suffix(self, taxid):
+        pass
+
```
I don't see this function used. Why is it defined?

dengzq1234 replied:
This is a function I want to do to handle GTDB accesssion,
because I realized sometime they use ids in both way, for example:
GCA_000003645.1	 GB_GCA_000003645.1
GCF_900113245.1   RS_GCF_900113245.1
they are equivalent so I want to have a function to handle the fuzzy. But I will keep it in my dev repo

jordibc replied:
Great, thanks!

jordibc commented on ete4/gtdb_taxonomy/gtdbquery.py:
```diff
@@ -146,13 +146,29 @@ def _translate_merged(self, all_taxids):
+    def _dirty_id_suffix(self, taxid):
+        pass
+
+    def _get_rank(self, taxids):
         """Return dictionary converting taxids to their GTDB taxonomy rank."""
         ids = ','.join('"%s"' % v for v in set(taxids) - {None, ''})
         result = self.db.execute('SELECT taxid, rank FROM species WHERE taxid IN (%s)' % ids)
         return {tax: spname for tax, spname in result.fetchall()}
+    
+    def get_rank(self, taxids):
+        taxid2rank = {}
+        name2ids = self._get_name_translator(taxids)
+        overlap_ids = name2ids.values()
+        taxids = [item for sublist in overlap_ids for item in sublist]
+        """Return dictionary converting taxids to their GTDB taxonomy rank."""
+        ids = ','.join('"%s"' % v for v in set(taxids) - {None, ''})
+        result = self.db.execute('SELECT taxid, rank FROM species WHERE taxid IN (%s)' % ids)
+        for tax, rank in result.fetchall():
+            taxid2rank[list(self._get_taxid_translator([tax]).values())[0]] = rank
+        
+        return taxid2rank
```
I don't understand this change. Why do we have `get_rank()` and `_get_rank()`. What's their difference? Also, `get_rank()` would need a docstring, and we should avoid having similarly-named functions since that makes it hard to guess which one is appropriate to use.

jordibc reviewed (changes requested):
Thanks @dengzq1234 

Before I merge this PR, I'd like to understand the couple of things I mentioned before, and also we should understand why after this change the ncbiquery test fails (which seems to be from your change in line 135 of `tests/test_ncbiquery.py`, that I imagine you had a reason to do):

```sh
$ pytest tests/test_ncbiquery.py 
[...]
tests/test_ncbiquery.py ......F.
[...]
____________________________________________ Test_ncbiquery.test_merged_id _____________________________________________

self = <tests.test_ncbiquery.Test_ncbiquery testMethod=test_merged_id>

    def test_merged_id(self):
      ncbi = NCBITaxa(dbfile=DATABASE_PATH)
      t1 = ncbi.get_lineage(649756)
>     self.assertEqual(t1, [1, 131567, 2, 1783272, 1239, 186801, 3085636, 186803, 207244, 649756])
E     AssertionError: Lists differ: [1, 131567, 2, 1783272, 1239, 186801, 186802, 186803, 207244, 649756] != [1, 131567, 2, 1783272, 1239, 186801, 3085636, 186803, 207244, 649756]
E     
E     First differing element 6:
E     186802
E     3085636
E     
E     - [1, 131567, 2, 1783272, 1239, 186801, 186802, 186803, 207244, 649756]
E     ?                                       ^  ^^^
E     
E     + [1, 131567, 2, 1783272, 1239, 186801, 3085636, 186803, 207244, 649756]
E     ?                                       ^^ + ^^

test_ncbiquery.py:135: AssertionError
```

dengzq1234 commented:
> Thanks @dengzq1234
> 
> Before I merge this PR, I'd like to understand the couple of things I mentioned before, and also we should understand why after this change the ncbiquery test fails (which seems to be from your change in line 135 of `tests/test_ncbiquery.py`, that I imagine you had a reason to do):
> 
> ```shell
> $ pytest tests/test_ncbiquery.py 
> [...]
> tests/test_ncbiquery.py ......F.
> [...]
> ____________________________________________ Test_ncbiquery.test_merged_id _____________________________________________
> 
> self = <tests.test_ncbiquery.Test_ncbiquery testMethod=test_merged_id>
> 
>     def test_merged_id(self):
>       ncbi = NCBITaxa(dbfile=DATABASE_PATH)
>       t1 = ncbi.get_lineage(649756)
> >     self.assertEqual(t1, [1, 131567, 2, 1783272, 1239, 186801, 3085636, 186803, 207244, 649756])
> E     AssertionError: Lists differ: [1, 131567, 2, 1783272, 1239, 186801, 186802, 186803, 207244, 649756] != [1, 131567, 2, 1783272, 1239, 186801, 3085636, 186803, 207244, 649756]
> E     
> E     First differing element 6:
> E     186802
> E     3085636
> E     
> E     - [1, 131567, 2, 1783272, 1239, 186801, 186802, 186803, 207244, 649756]
> E     ?                                       ^  ^^^
> E     
> E     + [1, 131567, 2, 1783272, 1239, 186801, 3085636, 186803, 207244, 649756]
> E     ?                                       ^^ + ^^
> 
> test_ncbiquery.py:135: AssertionError
> ```

I changed it because I updated my ncbi taxonomy database, I wonder if it would effect `test_ncbiquery.taxa.sqlite`? For my latest update of ncbi taxonomy, I won't pass this unitest if I don't change it.

dengzq1234 committed:
rename _get_rank to _get_id2rank

dengzq1234 committed:
Merge branch 'ete4' into ete4_#750

dengzq1234 committed:
update sytanx for new test

dengzq1234 commented:
@jordibc Already update with the next ncbi unitest

jordibc reviewed (approved):
Looks good, thanks @dengzq1234 !
